### DISCUSSION TITLE
simple paralellism with 2 or 4 GPUs

### DISCUSSION THREAD

**Discussion by jecampagne** [jax-ml/jax#21127] [Q&A]
Hello,
Looking at the doc on [shard_map](https://jax.readthedocs.io/en/latest/notebooks/shard_map.html), I'm quite lost...
If someone can convert this simple example which compute a 2D representation of a `f(x,y)` function 
using either 2 or 4 GPUs it would be help. Here is a 1-GPU (device) code using nested vmap.

```python
import jax
import jax.numpy as jnp
from jax import jit, vmap
from matplotlib import pyplot as plt

@jit
def func(x,y):
    return jnp.sin(jnp.sqrt(x**2+y**2))

Npix=2048
xvals = jnp.linspace(-2*np.pi,2*np.pi,Npix)
yvals = jnp.linspace(-2*np.pi,2*np.pi,Npix)
mapped = vmap(vmap(func, in_axes=(None, 0)), in_axes=(0, None))
result = mapped(xvals,yvals)

plt.figure(figsize=(5,5))
img = plt.imshow(result,cmap=mpl.cm.jet)
plt.show()
```
![image](https://github.com/google/jax/assets/20539759/16d1507c-5b7a-4a2f-8344-4f8c86b20718)

I had in mind to split the image in 2x1 or 2x2 grid sending to each device the computation of part of the figure
then glue the results to build the final image.
Thanks for your help.

**Comment by jecampagne** [comment]
No help?

**Comment by jakevdp** [comment]
The doc you link to is for more advanced usage – I'd suggest starting with [Introduction to sharded computation](https://jax.readthedocs.io/en/latest/sharded-computation.html).

**Reply by jecampagne** [reply]
Thanks. I will try and come back to share my code. Here it is:

```python
# my 2d functions
@jax.jit
def func(x,y):
    return jnp.sin(jnp.sqrt(x**2+y**2))

# where I want to sample to get a 2D image
Npix=2048
xvals = jnp.linspace(-2*jnp.pi,2*jnp.pi,Npix)
yvals = jnp.linspace(-2*jnp.pi,2*jnp.pi,Npix)
jax.debug.visualize_array_sharding(xvals)

# a numpy-like method using the vectorization of sqrt/sin
X,Y = jnp.meshgrid(xvals,yvals)
jax.debug.visualize_array_sharding(X)
img = func(X,Y)

# if one has a non smart vectorized func: usage of vmap (1)
mapped = jax.vmap(func, in_axes=(0, 0))
img0 = jax.jit(mapped)(X,Y)
print(jnp.allclose(img,img0))
jax.debug.visualize_array_sharding(img0)

# vmap (2) using directly the xvals, yvals vectors
mapped = jax.vmap(jax.vmap(func, in_axes=(None, 0)), in_axes=(0, None))
img1 = jax.jit(mapped)(xvals,yvals)
print(jnp.allclose(img,img1))
jax.debug.visualize_array_sharding(img1)

# using  4 GPUs: notice the usage of X,Y meshgrid
P = jax.sharding.PartitionSpec
devices = mesh_utils.create_device_mesh((2, 2))
mesh = jax.sharding.Mesh(devices, ('x', 'y'))
sharding = jax.sharding.NamedSharding(mesh, P('x', 'y'))
print(sharding)

img2 = shard_map(func, mesh=mesh, in_specs=P('x','y'), out_specs=P('x','y'))(X,Y)
print(jnp.allclose(img,img2))
jax.debug.visualize_array_sharding(img2)
```
First I have used numpy-like method of meshgrid using the intrinsic vectorization of the `sqr/sin` functions.
Then, using `vmap` one can use directly the 1D arrays `xvals`, `yvals`.  All computations were done in the GPU 0 device. 
Finally using the `sharding`methods one can see the 4 GPUs.

My questions then are then the following:
1) Does the img2 is really computed using the 4 GPUs or it is only the output that is sharded? 
2) Is there `shard_map` method like the nested `vmap` that would allow to use directly the 1D `xvals`, `yvals` w/o the duplication on the sampling locations?
3)  vocabulary: what is the difernce between 'data sharing' and 'data sharding'?

**Comment by jecampagne** [comment]
Hello, I'm back.
Let us define a 2d function that we want to sample to get an 2D image. 
In the following I have 4 GPUs

```python
@jax.jit
def func(x,y):
    return jnp.sin(jnp.sqrt(x**2+y**2))

Npix=2048
xvals = jnp.linspace(-2*jnp.pi,2*jnp.pi,Npix)
yvals = jnp.linspace(-2*jnp.pi,2*jnp.pi,Npix)
```
`xvals` (and `yvals`) are in the GPU 0
```python
jax.debug.visualize_array_sharding(xvals)
```
One numpy-like away to proceed to the sampling is the define
```python
X,Y = jnp.meshgrid(xvals,yvals)
```
(`X` and `Y`are located in the GPU 0 as well, and there is a duplication of the sampling locations but see later).
So, to get the image one would certainly use the intrinsic vectorization of the jnp.sqr/jnp.sin functions and use
```python
img = func(X,Y)
```
and nicely plot the image as I have show in the first post. So far so good.

Let then first assume that my function is not so smart, and I cannot directly use `func(X,Y)`.  Then, I can use `jax.vmap`.
One   possibility is the following:
```python
mapped = jax.vmap(func, in_axes=(0, 0))
img0 = jax.jit(mapped)(X,Y)
jnp.allclose(img,img0)
jax.debug.visualize_array_sharding(img0)
```
and one can see that img0 is computed in the GPU 0 device. 

Now, one can question the usefulness of the  duplication of the sampling locations, 
and we can directly use the `xvals, yvals`arrays as follows:
```python
mapped = jax.vmap(jax.vmap(func, in_axes=(None, 0)), in_axes=(0, None))
img1 = jax.jit(mapped)(xvals,yvals)
print(jnp.allclose(img0,img1))
jax.debug.visualize_array_sharding(img1)
```

Now, all the above computations use a single device. Let us investigate my original request on how I can use my 4 GPUs?
Here is what I have done, notice btw that I'm using the X,Y `jnp.meshgrid` done above:
```python
P = jax.sharding.PartitionSpec
devices = mesh_utils.create_device_mesh((2, 2))
mesh = jax.sharding.Mesh(devices, ('x', 'y'))
sharding = jax.sharding.NamedSharding(mesh, P('x', 'y'))
img2 = shard_map(func, mesh=mesh, in_specs=P('x','y'), out_specs=P('x','y'))(X,Y)
print(jnp.allclose(img,img2))
jax.debug.visualize_array_sharding(img2)
```
First it gives the correct output and the computation is done with my 4 GPus
![image](https://github.com/google/jax/assets/20539759/1480f5ad-a90f-475d-aed1-b95eb9177121)

Now, I would like to see if I'm right to say that the compuation is done using my 4 GPUs and not just the output is shared among the 4 devices? 
And the second question is: is there a way to use only the xvals, yvals arrays and not the X,Y meshgrid?

**Reply by jakevdp** [reply]
> I'm right to say that the compuation is done using my 4 GPUs and not just the output is shared among the 4 devices?

Yes, the computation will be sharded across the 4 GPUs in this case: the `in_specs` mean that your data is sharded across the devices in this way, and computation will follow data.

> is there a way to use only the xvals, yvals arrays and not the X,Y meshgrid?

Yes – for example, you could wrap your `mapped` function in `shard_map` like this:
```python
mapped = jax.vmap(jax.vmap(func, in_axes=(None, 0)), in_axes=(0, None))
img2 = shard_map(mapped, mesh=mesh, in_specs=(P('x'), P('y')), out_specs=P('x','y'))(xvals, yvals)
```

**Reply by jecampagne** [reply]
Great! so the `shard_map` replace the jit? what would the "Automatic parallelism via jit" option using my 4 GPUs?

**Reply by jakevdp** [reply]
No, `shard_map` doesn't replace `jit` – you could wrap `jit` around the outside of the `shard_map` to improve your performance.

For the automatic parallelism approach, you could shard the input arrays, and then pass them to the jit-compiled `mapped` function:
```python
x_sharded = jax.device_put(xvals, jax.sharding.NamedSharding(mesh, P('x')))
y_sharded = jax.device_put(yvals, jax.sharding.NamedSharding(mesh, P('y')))
img = jax.jit(mapped)(x_sharded, y_sharded)
```
You should see that the output sharding on `img` is the same as in the shard map case – it's just determined automatically by the compiler to be the optimal sharding for this computation given the sharding of the inputs.

**Reply by jecampagne** [reply]
Okay. thanks a lot. I will try all that on a more complicated use case.

**Reply by jecampagne** [reply]
Ho, I think to a new question. Imagine that my entire 2D image computation does not fit in a single GPU; but 
looking at 1/4 of the (x,y) it is ok on a sinhgle GPU, so I can compute the entire 2D image using the 4 GPUs

Does one of the above soultion is ok, and which one?

**Reply by jakevdp** [reply]
Any of those solutions should work in that case: you can make sure the data is sharded across the devices where you want the computation to take place.
